packages/flutter_tools/lib/src/base/net.dart

```diff
@@ -70,7 +70,7 @@ class Net {
 
       if (maxAttempts != null && attempts >= maxAttempts) {
         _logger.printStatus('Download failed -- retry $attempts');
-        return null;
+        throw Exception('Failed to download $url');
```

Comment:
`_attempt` could return `null` on success, and the exception object on an error. Then here you can rethrow the most recent error. Then at the callsite, you can catch `IOException`, or something more narrow, since it doesn't look like the text `'Failed to download...'` is used there. The caller also presumably already knows what the url is.

Reply:
done?